### ISSUE TITLE
torch.scatter_logsumexp

### ISSUE THREAD

**Issue by fuzic** [pytorch/pytorch#31394]
## 🚀 Feature
torch.scatter_add will distribute values over an output tensor, summing if multiple values land in the same destination coordinate.

torch.logsumexp performs addition in linear space of quantities with log-space values, useful for adding probabilities which are stored as log likelihoods, for example.

torch.scatter_logsumexp would behave the same as torch.scatter_add, but instead of linear summing, it would logsumexp to sum.

## Motivation

I am experimenting with alterations to CTC, and python implementations of CTC have to sum probabilities from identical labels in a target sequence to get the total probability of a particular token. The algorithm is much more numerically stable if implemented with log likelihoods instead of linear probabilities, and torch.scatter_logsumexp would simplify this kind of thing.

## Alternatives

Currently either log likelihoods have to be exponential'ed and then scatter_add'ed (numerically unstable), or scattered into a target tensor with an expanded extra dimension (so no two values land together) and then logsumexp'd over that extra dimension

**Comment by cpuhrsch** [comment]
We're taking this into account as an enhancement - we're willing to accept a PR

**Comment by cpuhrsch** [comment]
The wider issue discussing this is https://github.com/pytorch/pytorch/issues/22378

**Comment by ngimel** [comment]
there's a package https://github.com/rusty1s/pytorch_scatter that provides this functionality.

**Comment by jeanmonet** [comment]
> there's a package https://github.com/rusty1s/pytorch_scatter that provides this functionality.

AFAIK that package is being decommissioned in favor if scatter methods being directly integrated into pytorch.

**Comment by alexshtf** [comment]
I think it's extremely useful for things like contrastive loss, and learning to rank, where the number of samples in the log-sum-exp expression may vary per query, and thus the computation cannot be performed by a log-sum-exp over one of the axes of a matrix.

**Comment by tvercaut** [comment]
Such a feature would be very useful, including for reduction of sparse matrices. We implemented a (python-only) workaround for this latter specific use case in [`torchsparsegradutils`](https://github.com/cai4cai/torchsparsegradutils/blob/main/torchsparsegradutils/sparse_logsumexp.py) but an upstream solution would be much better. Similar to https://github.com/pytorch/pytorch/issues/31837#issuecomment-3140857800 performance could be significantly improved by implementing a one-pass / streaming algorithm, à la [Nowozin](https://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html).
